lib/beacon/compiler.ex

```diff
@@ -0,0 +1,125 @@
+  @moduledoc false
+
+  require Logger
+  alias Beacon.Loader
+
+  @type diagnostics :: [Code.diagnostic(:warning | :error)]
+
+  @spec compile_module(Beacon.Site.t(), Macro.t(), String.t()) ::
+          {:ok, module(), diagnostics()} | {:error, module(), {Exception.t(), diagnostics()}} | {:error, Exception.t() | :invalid_module}
+  def compile_module(site, quoted, file \\ "nofile") do
+    case module_name(quoted) do
+      {:ok, module} ->
+        do_compile_module(site, module, quoted, hash(quoted), file)
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  defp do_compile_module(site, module, quoted, hash, file) do
+    case {:erlang.module_loaded(module), current_hash(site, module) == hash} do
+      {true, true} ->
+        {:ok, module, []}
+
+      {true, _} ->
+        unload(module)
+        compile_and_register(site, module, quoted, hash, file)
+
+      {false, _} ->
+        compile_and_register(site, module, quoted, hash, file)
+    end
+  end
+
+  def module_name({:defmodule, _, [{:__aliases__, _, [module]}, _]}) do
+    {:ok, Module.concat([module])}
+  end
+
+  def module_name({:defmodule, _, [module, _]}) do
+    {:ok, Module.concat([module])}
+  end
+
+  def module_name({module, {:defmodule, _, _}}) do
+    {:ok, Module.concat([module])}
+  end
+
+  def module_name(quoted) do
+    Logger.error("""
+    invalid module given to Beacon.Compiler,
+    expected a quoted expression containing a single module.
+
+      Got: #{inspect(quoted)}
+
+    """)
+
+    {:error, :invalid_module}
+  end
+
+  defp compile_and_register(site, module, quoted, hash, file) do
+    case compile_quoted(quoted, file) do
+      {:ok, module, diagnostics} ->
+        add_module(site, module, hash, nil, diagnostics)
+        {:ok, module, diagnostics}
+
+      {:error, error, diagnostics} ->
+        add_module(site, module, hash, error, diagnostics)
+        {:error, module, {error, diagnostics}}
+
+      {:error, error} ->
+        add_module(site, module, hash, error, nil)
+        {:error, error}
+    end
+  end
+
+  if Version.match?(System.version(), ">= 1.15.0") do
+    defp compile_quoted(quoted, file) do
+      {result, diagnostics} =
+        Code.with_diagnostics(fn ->
+          try do
+            [{module, _}] = Code.compile_quoted(quoted, file)
+            {:module, ^module} = Code.ensure_loaded(module)
+            {:ok, module}
+          rescue
+            error -> {:error, error}
+          end
+        end)
+
+      case result do
+        {:ok, module} ->
+          {:ok, module, diagnostics}
+
+        {:error, error} ->
+          {:error, error, diagnostics}
+      end
+    end
+  else
+    defp compile_quoted(quoted, file) do
+      [{module, _}] = Code.compile_quoted(quoted, file)
+      {:module, ^module} = Code.ensure_loaded(module)
+      {:ok, module, []}
+    rescue
+      error -> {:error, error, []}
+    end
+  end
+
+  def unload(module) do
+    :code.purge(module)
+    :code.delete(module)
+  end
```

Comment:
I've been wondering about the order of these two calls, ever since I tried running some benchmarks last week and reached an error which seemed to be due to these being called in the wrong order.

Looking at the docs, it says that when calling `:code.delete/1` "the current code for Module is made old"

Then `:code.purge/1` "removes code marked as old"

so it seems like it should be delete first, then purge.

There is this note in `:code.delete/1` "Returns [...] false if there is old code for Module that must be purged first" but I think this just means you can't call `:code.delete/1` twice in a row and have TWO modules worth of "old" code.  If we always call the two together, we shouldn't be affected by that.

Reply:
It's basically swapping the current module for the new one being compiled, think of it like pointers. It works similar to mix task compile or phoenix code reloader.

First it removes (purge) the old module to make room for another one, so calling delete will move the current to old, and then compile will place the new module into current.

That works because we know a new module will be placed into current, ending up with old and current in the right spots.

So the workflow for a module `A` that gets updated from v2 to v3 would look like:

- `purge` v1 from old
- move v2 to old (delete)
- place v3 as current (compile)

And the workflow from v1 to v2, ie: in this case there's no old:

- `purge` nothing from old (returns false because there's only current v1 loaded)
- move v1 to old (delete)
- place v2 as current (compile)

It's important to keep the "previous version" as old because some process may still be using that module.

Reply:
`delete` + `purge` is useful when you want to get rid of that module completely, so you delete first to move it to old then purge to remove the old version, ending up with nothing, the module is no longer available in any way.

Reply:
Ah, thanks so much for detailing these workflows and with your second comment I now see the difference between Beacon (keeping old code for concurrent processes) and my other case (no concurrent processes) 👍